compiler/rustc_infer/src/infer/mod.rs

```diff
@@ -687,15 +687,25 @@ pub struct CombinedSnapshot<'a, 'tcx> {
 impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
     /// calls `tcx.try_unify_abstract_consts` after
     /// canonicalizing the consts.
+    #[instrument(skip(self), level = "debug")]
     pub fn try_unify_abstract_consts(
         &self,
         a: ty::Unevaluated<'tcx, ()>,
         b: ty::Unevaluated<'tcx, ()>,
+        param_env: ty::ParamEnv<'tcx>,
     ) -> bool {
+        // Reject any attempt to unify two unevaluated constants that contain inference
+        // variables.
+        // FIXME `TyCtxt::const_eval_resolve` already rejects the resolution of those
+        // constants early, but the canonicalization below messes with that mechanism.
+        if a.substs.has_infer_types_or_consts() || b.substs.has_infer_types_or_consts() {
+            debug!("a or b contain infer vars in its substs -> cannot unify");
+            return false;
+        }
+
         let canonical = self.canonicalize_query((a, b), &mut OriginalQueryValues::default());
```

Comment:
can remove that, canonicalization replaces inference vars with bound vars.

Reply:
No, this is actually necessary. If we don't reject these here we get the `delay_span_bug` ICEs in `resolve_opt_const_arg` because the normalization fails with the bound vars.

Reply:
I meant remove the call to `canonicalize_query`. That one is a noop if we do not get here with infer vars.

You have to replace it with `tcx.erase_regions` however

Reply:
you might also have to canonicalize the `param_env` here, we sometimes have inference vars in that as well